File: app/src/main/java/com/zhou/android/main/CameraH264Activity.java
package com.zhou.android.main;

import android.Manifest;
import android.content.Context;
import android.content.pm.PackageManager;
import android.graphics.ImageFormat;
import android.graphics.SurfaceTexture;
import android.hardware.camera2.CameraAccessException;
import android.hardware.camera2.CameraCaptureSession;
import android.hardware.camera2.CameraCharacteristics;
import android.hardware.camera2.CameraDevice;
import android.hardware.camera2.CameraManager;
import android.hardware.camera2.CaptureFailure;
import android.hardware.camera2.CaptureRequest;
import android.hardware.camera2.TotalCaptureResult;
import android.hardware.camera2.params.StreamConfigurationMap;
import android.media.Image;
import android.media.ImageReader;
import android.os.Build;
import android.os.Handler;
import android.support.annotation.NonNull;
import android.support.annotation.RequiresApi;
import android.support.v4.app.ActivityCompat;
import android.util.Log;
import android.util.Size;
import android.view.Surface;
import android.view.TextureView;
import android.view.View;
import android.view.ViewGroup;

import com.zhou.android.R;
import com.zhou.android.common.BaseActivity;
import com.zhou.android.common.ToastUtils;
import com.zhou.android.common.Utils;

import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;

import io.reactivex.disposables.Disposable;

/**
 * 双摄（Camera2 API）
 * Created by Administrator on 2018/10/10.
 */
@RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
public class CameraH264Activity extends BaseActivity {

    private final static String FRONT = "camera_front";
    private final static String BACK = "camera_back";

    private CameraManager cameraManager;
    private TextureView textureBack;
    private Size cameraSize;
    private int degree = 0;

    private Disposable disposable;

    private HashMap<String, CameraConfig> map = new HashMap<>();
    private boolean isBackCamera = true;

    private ImageReader imageReader;
    private ImageReader.OnImageAvailableListener onImageAvailableListener = new ImageReader.OnImageAvailableListener() {
        @Override
        public void onImageAvailable(ImageReader reader) {
            Image image = reader.acquireNextImage();
            Log.d("image", "size [ w: " + image.getWidth() + " h: " + image.getHeight() + " ]");
        }
    };

    @Override
    protected void setContentView() {
        setContentView(R.layout.activity_camera_h264);
    }

    @Override
    protected void init() {
        textureBack = findViewById(R.id.texture_back);

        Handler backgroundHandler = new Handler();

        cameraManager = (CameraManager) getSystemService(Context.CAMERA_SERVICE);
        try {
            String[] ids = cameraManager.getCameraIdList();
            for (String id : ids) {
                CameraCharacteristics characteristics = cameraManager.getCameraCharacteristics(id);
                Integer cameraOrientation = characteristics.get(CameraCharacteristics.LENS_FACING);
                if (cameraOrientation == null)
                    continue;
                if (cameraOrientation == CameraCharacteristics.LENS_FACING_BACK) {
                    map.put(BACK, new CameraConfig(id, characteristics, deviceBackStateCallback, backgroundHandler));
                    StreamConfigurationMap size = characteristics.get(CameraCharacteristics.SCALER_STREAM_CONFIGURATION_MAP);
                    if (size != null) {
                        Size largest = Collections.max(Arrays.asList(size.getOutputSizes(ImageFormat.JPEG)),
                                new Comparator<Size>() {
                                    @Override
                                    public int compare(Size lhs, Size rhs) {
                                        return Long.signum((long) lhs.getWidth() * lhs.getHeight() -
                                                (long) rhs.getWidth() * rhs.getHeight());
                                    }
                                });
                        cameraSize = largest;
                        Integer degree = characteristics.get(CameraCharacteristics.SENSOR_ORIENTATION);
                        if (degree != null) {
                            this.degree = degree;
                        }

                        imageReader = ImageReader.newInstance(largest.getWidth(), largest.getHeight(), ImageFormat.YUV_420_888, 2);
                        imageReader.setOnImageAvailableListener(onImageAvailableListener, backgroundHandler);
                    }
                }
            }
        } catch (CameraAccessException e) {
            e.printStackTrace();
        }

        disposable = Utils.checkPermission(this, Manifest.permission.CAMERA, "相机");

    }

    @Override
    protected void addListener() {

    }

    @Override
    protected void onResume() {
        super.onResume();
        if (textureBack.isAvailable()) {
            openCamera(map.get(BACK));
        } else {
            textureBack.setSurfaceTextureListener(textureListenerBack);
        }
    }

    private TextureView.SurfaceTextureListener textureListenerBack = new TextureView.SurfaceTextureListener() {
        @Override
        public void onSurfaceTextureAvailable(SurfaceTexture surface, int width, int height) {
            int w = cameraSize.getWidth();
            int h = cameraSize.getHeight();
            if (degree == 0 || degree == 180) {
                w = cameraSize.getHeight();
                h = cameraSize.getWidth();
            }

            ViewGroup.LayoutParams lp = textureBack.getLayoutParams();
            lp.height = (int) (textureBack.getWidth() * 1.0f * w / h);
            textureBack.setLayoutParams(lp);
            textureBack.requestLayout();

            openCamera(map.get(BACK));
        }

        @Override
        public void onSurfaceTextureSizeChanged(SurfaceTexture surface, int width, int height) {

        }

        @Override
        public boolean onSurfaceTextureDestroyed(SurfaceTexture surface) {
            return true;
        }

        @Override
        public void onSurfaceTextureUpdated(SurfaceTexture surface) {

        }
    };

    private CameraDevice.StateCallback deviceBackStateCallback = new CameraDevice.StateCallback() {
        @Override
        public void onOpened(@NonNull CameraDevice camera) {
            createCameraSession(textureBack, map.get(BACK), camera);
        }

        @Override
        public void onDisconnected(@NonNull CameraDevice camera) {
            camera.close();
            Log.d("camera", "disconnect");
        }

        @Override
        public void onError(@NonNull CameraDevice camera, int error) {
            camera.close();
            Log.d("camera", "error " + error);
        }
    };

    private CameraCaptureSession.CaptureCallback captureCallback = new CameraCaptureSession.CaptureCallback() {

        @Override
        public void onCaptureCompleted(@NonNull CameraCaptureSession session, @NonNull CaptureRequest request, @NonNull TotalCaptureResult result) {
            super.onCaptureCompleted(session, request, result);
        }

        @Override
        public void onCaptureFailed(@NonNull CameraCaptureSession session, @NonNull CaptureRequest request, @NonNull CaptureFailure failure) {
            super.onCaptureFailed(session, request, failure);
        }
    };

    private void openCamera(CameraConfig config) {
        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
            ToastUtils.show(this, "This app does not have camera permission");
            return;
        }
        if (config == null) {
            Log.d("camera", "CameraConfig is null.");
            return;
        }
        String cameraId = config.cameraId;
        if (cameraManager == null) {
            cameraManager = (CameraManager) getSystemService(Context.CAMERA_SERVICE);
        }
        try {
            cameraManager.openCamera(cameraId, config.deviceStateCallback, config.handler);
        } catch (CameraAccessException e) {
            e.printStackTrace();
        }
    }

    private void createCameraSession(TextureView textureView, final CameraConfig config, CameraDevice cameraDevice) {

        Surface surface = new Surface(textureView.getSurfaceTexture());
        try {
            if (config != null) {
                config.cameraDevice = cameraDevice;
            }
            final CaptureRequest.Builder requestBuilder = cameraDevice.createCaptureRequest(CameraDevice.TEMPLATE_PREVIEW);
            requestBuilder.addTarget(surface);

            cameraDevice.createCaptureSession(Arrays.asList(surface), new CameraCaptureSession.StateCallback() {
                @Override
                public void onConfigured(@NonNull CameraCaptureSession session) {

                    if (null != config) {
                        config.cameraCaptureSession = session;
                        requestBuilder.set(CaptureRequest.CONTROL_AF_MODE, CaptureRequest.CONTROL_AF_MODE_CONTINUOUS_PICTURE);

                        try {
                            session.setRepeatingRequest(requestBuilder.build(), captureCallback, config.handler);
                        } catch (CameraAccessException e) {
                            e.printStackTrace();
                        }
                    }
                }

                @Override
                public void onConfigureFailed(@NonNull CameraCaptureSession session) {
                    Log.d("camera", "config failed.");
                }
            }, null);

        } catch (CameraAccessException e) {
            e.printStackTrace();
        }
    }

    private void closeCamera(CameraConfig config) {
        if (config != null) {
            if (config.cameraCaptureSession != null) {
                config.cameraCaptureSession.close();
                config.cameraCaptureSession = null;
            }
            if (config.cameraDevice != null) {
                config.cameraDevice.close();
                config.cameraDevice = null;
            }
        }
    }

    private class CameraConfig {

        String cameraId;
        int orientation;
        boolean fullSupport = false;
        CameraDevice cameraDevice;
        CameraDevice.StateCallback deviceStateCallback;
        CameraCaptureSession cameraCaptureSession;

        Handler handler;

        CameraConfig(String cameraId, CameraCharacteristics characteristics, @NonNull CameraDevice.StateCallback deviceStateCallback, @NonNull Handler handler) {
            this.cameraId = cameraId;
            Integer orientation = characteristics.get(CameraCharacteristics.SENSOR_ORIENTATION);
            this.orientation = orientation == null ? 0 : orientation;
            Integer level = characteristics.get(CameraCharacteristics.INFO_SUPPORTED_HARDWARE_LEVEL);
            fullSupport = level != null && level == CameraCharacteristics.INFO_SUPPORTED_HARDWARE_LEVEL_FULL;
            this.deviceStateCallback = deviceStateCallback;
            this.handler = handler;
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (disposable != null && !disposable.isDisposed()) {
            disposable.dispose();
            disposable = null;
        }
        closeCamera(map.get(isBackCamera ? BACK : FRONT));
        if (imageReader != null) {
            imageReader.close();
            imageReader = null;
        }
    }
}

File: app/src/main/java/com/zhou/android/main/CameraH264Activity.kt
package com.zhou.android.main

import android.Manifest
import android.annotation.SuppressLint
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.graphics.ImageFormat
import android.graphics.Rect
import android.graphics.YuvImage
import android.media.ImageReader
import android.os.*
import android.support.annotation.RequiresApi
import android.support.v4.app.ActivityCompat
import android.support.v7.app.AppCompatActivity
import android.util.Log
import com.squareup.picasso.MemoryPolicy
import com.squareup.picasso.Picasso
import com.tbruyelle.rxpermissions2.RxPermissions
import com.wuwang.libyuv.Key
import com.wuwang.libyuv.YuvUtils
import com.zhou.android.R
import com.zhou.android.camera2.CameraUtil
import com.zhou.android.camera2.OnImageAvailableListener
import com.zhou.android.common.Utils
import com.zhou.android.common.addToComposite
import com.zhou.android.common.toast
import io.reactivex.Observable
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.activity_camera_h264.*
import okhttp3.internal.Util
import java.io.*

/**
 * Created by mxz on 2019/10/31.
 */
@RequiresApi(Build.VERSION_CODES.LOLLIPOP)
class CameraH264Activity : AppCompatActivity() {

    private val composite = CompositeDisposable()
    private var cameraUtil: CameraUtil? = null

    private var hasCameraPermission = false

    private val path = "${Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES).path}/android/"
    @Volatile
    private var takePic = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_camera_h264)
        init()
        addListener()
    }

    private fun init() {

        if (PackageManager.PERMISSION_GRANTED == ActivityCompat.checkSelfPermission(this, Manifest.permission.CAMERA)) {
            try {
                initCameraUtil()
            } catch (e: IllegalArgumentException) {
                toast("不支持的编码格式")
            }
        }

    }

    private fun addListener() {
        ibRecord.setOnClickListener {
            val state = !ibRecord.isSelected
            if (state) {
                toast("开始录屏")
            } else {
                toast("停止录屏")
            }

            ibRecord.isSelected = state
        }

        ibVideo.setOnClickListener {
            toast("视频列表")
        }

        btnShot.setOnClickListener {
            if (PackageManager.PERMISSION_GRANTED == ActivityCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE)) {
                takePic = true
            } else {
                val context = this@CameraH264Activity
                val rxPermissions = RxPermissions(context)
                rxPermissions.shouldShowRequestPermissionRationale(context, Manifest.permission.WRITE_EXTERNAL_STORAGE)
                        .flatMap { b ->
                            if (b) {
                                rxPermissions.request(Manifest.permission.WRITE_EXTERNAL_STORAGE)
                            } else {
                                Observable.just(false)
                            }
                        }.subscribe { result ->
                            if (!result) {
                                Utils.showPermissionRequestAlertDialog(context, "存储")
                            }
                        }.addToComposite(composite)
            }
        }

    }

    override fun onResume() {
        super.onResume()
        val rxPermissions = RxPermissions(this)
        hasCameraPermission = rxPermissions.isGranted(Manifest.permission.CAMERA)
        if (!hasCameraPermission) {
            rxPermissions.shouldShowRequestPermissionRationale(this, Manifest.permission.CAMERA)
                    .flatMap { r ->
                        if (r) {
                            rxPermissions.request(Manifest.permission.CAMERA)
                        } else {
                            Observable.just(false)
                        }
                    }.subscribe { result ->
                        if (!result) {
                            Utils.showPermissionRequestAlertDialog(this@CameraH264Activity, "相机")
                        } else {
                            initCameraUtil()
                        }
                    }.addToComposite(composite)
        }

        cameraUtil?.startPreview(this)
    }

    override fun onPause() {
        cameraUtil?.stopPreview()
        super.onPause()
    }

    override fun onDestroy() {
        super.onDestroy()
        composite.dispose()
    }

    private fun initCameraUtil() {
        if (cameraUtil == null) {
            cameraUtil = CameraUtil(this, null, texture).apply {
                setDefaultCamera(true)
                setPreviewImageFormat(ImageFormat.YUV_420_888)
                setOnImageAvailableListener(onImageAvailableListener)
//                setSurfaceCallback { config ->
//                    val size = config.size
//                    Log.d("zhou1234", "SurfaceCallback >> w:${size.width} ,h:${size.height}")
////                    宽高要反过来算
//                    val lp = texture.layoutParams
//                    lp.height = (texture.width * 1.0f * size.width / size.height).toInt()
//                    texture.layoutParams = lp
//                    texture.requestLayout()
//                }
            }
        }
    }

    private val sb = StringBuilder()
    private val smallBuf = ByteArray(10)
    private val onImageAvailableListener = object : OnImageAvailableListener() {

        override fun onImageAvailable(reader: ImageReader) {

            val image = reader.acquireLatestImage()

            if (takePic) {
                //这里是 YUV420SemiPlanar yyyyuvuv 通道二三数据一致，只是二通道是 uv 间隔，三通道是 vu 间隔
                sb.setLength(0)
                sb.append("width:${image.width}, height:${image.height} ,format:${image.format} \n")
                val data = ByteArray(image.width * image.height * 3 / 2)
                var count = 0
                for (i in 0 until image.planes.size - 1) {
                    val plane = image.planes[i]

                    val buf = plane.buffer
                    val remaining = buf.remaining()
                    val b = ByteArray(remaining)
                    buf.get(b)

                    System.arraycopy(b, 0, data, count, b.size)

                    count += remaining - 1

//                    System.arraycopy(b, 0, smallBuf, 0, 10)
//                    sb.append("plane[$i], size:$remaining ,stride:${plane.pixelStride} ,10 byte >> ")
//                            .append(Arrays.toString(smallBuf)).append("\n")
                }
                Log.d("zhou1234", sb.append("=========================\n").toString())

                createBitmap(data, image.width, image.height)

                takePic = false
            }
            image.close()
        }
    }

    private fun createBitmap(yuv: ByteArray, width: Int, height: Int) {

        Single.fromCallable {

            //原数据是NV21 yuv420sp，如果有角度需旋转
            //使用的库是 https://github.com/doggycoder/AndroidLibyuv.git 在这基础上做了修改
            //也可以用自带的 NV21ToI420 旋转为i420，再将i420转NV21 I420ToNV21
            //注：i420 的编码是 yyyy uu vv ，NV21是 yyyy vuvu ，后面参数需要传true，宽高需调换
            // I420ToNV21(src,dst,h,w,true)
            var w = width
            var h = height

            val degree = cameraUtil?.sensorOrientation ?: 0
            val dest = if (degree == 0) {
                yuv
            } else {
                val time = System.currentTimeMillis()
                val trans = ByteArray(yuv.size)
                val de = when (degree) {
                    90 -> {
                        w = height
                        h = width
                        Key.ROTATE_90
                    }
                    180 -> Key.ROTATE_180
                    270 -> {
                        w = height
                        h = width
                        Key.ROTATE_270
                    }
                    else -> Key.ROTATE_90
                }
                YuvUtils.NV21Rotate(yuv, width, height, trans, de, false)
                Log.d("zhou1234", "NV21Rotate time = ${System.currentTimeMillis() - time}ms")
                trans
            }

            val image = YuvImage(dest, ImageFormat.NV21, w, h, null)
            val bos = ByteArrayOutputStream()
            image.compressToJpeg(Rect(0, 0, w, h), 100, bos)
            val array = bos.toByteArray()

            val file = File("$path${System.currentTimeMillis()}.jpg")
            var fos: BufferedOutputStream? = null
            try {
                if (!file.parentFile.exists()) {
                    file.parentFile.mkdirs()
                }
                file.createNewFile()
                fos = BufferedOutputStream(FileOutputStream(file))
                fos.write(array, 0, array.size)
                fos.flush()
                fos.close()
            } catch (e: IOException) {
                e.printStackTrace()
                Log.e("zhou1234", "error > ${e.message}")
            } finally {
                Util.closeQuietly(fos)
                Util.closeQuietly(bos)
            }
            file
        }.subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe { file ->
                    Picasso.with(this@CameraH264Activity)
                            .load(file)
                            .config(Bitmap.Config.RGB_565)
                            .priority(Picasso.Priority.LOW)
                            .memoryPolicy(MemoryPolicy.NO_CACHE)
                            .into(image)
                }.addToComposite(composite)

    }

    private val handler = @SuppressLint("HandlerLeak")
    object : Handler() {
        override fun handleMessage(msg: Message) {
            super.handleMessage(msg)
            val bundle = msg.data
            val buf = bundle.getByteArray("data")
            val width = bundle.getInt("width")
            val height = bundle.getInt("height")
            createBitmap(buf, width, height)
        }
    }
}